### PR TITLE
[GraphBolt] Fix scalar itemset dtype issue

### PR DESCRIPTION
**Pull request by mfbalin**
## Description
Scalar torch tensor option is added to ItemSet so that we can generate `all_nodes_set` with a specific dtype.

Should help #7127 CI tests to pass.

Also related to #7130.

Now, we will need to modify all the itemsets in our BuiltinDatasets that are simple integers and wrap them into a `torch.tensor(N, dtype=dtype)`.

## Checklist
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [ ] I've leverage the [tools](https://docs.google.com/document/d/1iHyj7zlmygKSk5gBPsqIqL5ASPzJSPREaNT_QdsiYA4/edit) to beautify the python and c++ code.
- [ ] The PR is complete and small, read the [Google eng practice (CL equals to PR)](https://google.github.io/eng-practices/review/developer/small-cls.html) to understand more about small PR. In DGL, we consider PRs with less than 200 lines of core code change are small (example, test and documentation could be exempted).
- [ ] All changes have test coverage
- [ ] Code is well-documented
- [ ] To the best of my knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR
- [ ] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

### CODE REVIEW

**Review by frozenbugs — APPROVED** [review]

**Review by Skeleton003 — APPROVED** [review]

**Comment by Skeleton003** [comment]
@mfbalin please take a look at https://github.com/dmlc/dgl/blob/master/python/dgl/graphbolt/impl/ondisk_dataset.py#L824.
OnDiskDataset's `_init_all_nodes_set` makes use of `num_nodes` API of `FusedCSCSamplingGraph` to generate `all_nodes_set`. Could you please also modify the return type of `FusedCSCSamplingGraph.num_nodes` from `Union[int, Dict[str, int]]` to `Union[torch.Tensor, Dict[str, torch.Tensor]]` and test if your changes work correctly? Many thanks.

**Comment by frozenbugs** [comment]
> @mfbalin please take a look at https://github.com/dmlc/dgl/blob/master/python/dgl/graphbolt/impl/ondisk_dataset.py#L824. OnDiskDataset's `_init_all_nodes_set` makes use of `num_nodes` API of `FusedCSCSamplingGraph` to generate `all_nodes_set`. Could you please also modify the return type of `FusedCSCSamplingGraph.num_nodes` from `Union[int, Dict[str, int]]` to `Union[torch.Tensor, Dict[str, torch.Tensor]]` and test if your changes work correctly? Many thanks.

I don't think we need to change `num_nodes` return type to Tensor, int should be fine.

**Comment by Skeleton003** [comment]
> > @mfbalin please take a look at https://github.com/dmlc/dgl/blob/master/python/dgl/graphbolt/impl/ondisk_dataset.py#L824. OnDiskDataset's `_init_all_nodes_set` makes use of `num_nodes` API of `FusedCSCSamplingGraph` to generate `all_nodes_set`. Could you please also modify the return type of `FusedCSCSamplingGraph.num_nodes` from `Union[int, Dict[str, int]]` to `Union[torch.Tensor, Dict[str, torch.Tensor]]` and test if your changes work correctly? Many thanks.
> 
> I don't think we need to change `num_nodes` return type to Tensor, int should be fine.

You're right. So what is supposed to be modified is `OnDiskDataset._init_all_nodes_set`.

File: python/dgl/graphbolt/itemset.py [hunk 1]
```diff
@@ -97,7 +117,7 @@ def __init__(
         items: Union[int, Iterable, Tuple[Iterable]],
```
Comment (Skeleton003) [inline python/dgl/graphbolt/itemset.py:117]:
```suggestion
        items: Union[int, torch.Tensor, Iterable, Tuple[Iterable]],
```
Is it necessary to specify that torch.Tensor(scalar) is supported?

**Comment by frozenbugs** [comment]
> _init_all_nodes_set

Why do we need to change _init_all_nodes_set?

**Comment by Skeleton003** [comment]
> > _init_all_nodes_set
> 
> Why do we need to change _init_all_nodes_set?

Please take a kind look at the [Description](https://github.com/dmlc/dgl/pull/7147#issue-2151742994). The purpose of this PR is to generate `all_nodes_set` with a specific dtype, and this purpose cannot be achieved if neither `FusedCSCSamplingGraph.num_nodes` nor `OnDiskDataset.all_nodes_set` is changed because currently `FusedCSCSamplingGraph.num_nodes` always returns `Union[int, Dict[str, int]]` and `ItemSet` cannot judge the dtype of a `int`.

**Comment by frozenbugs** [comment]
> > > _init_all_nodes_set
> > 
> > 
> > Why do we need to change _init_all_nodes_set?
> 
> Please take a kind look at the [Description](https://github.com/dmlc/dgl/pull/7147#issue-2151742994). The purpose of this PR is to generate `all_nodes_set` with a specific dtype, and this purpose cannot be achieved if neither `FusedCSCSamplingGraph.num_nodes` nor `OnDiskDataset._init_all_nodes_set` is changed, because currently `FusedCSCSamplingGraph.num_nodes` always returns `Union[int, Dict[str, int]]` and `ItemSet` cannot judge the dtype of an `int`.

How the dtype should be decided? By external users or depend on any attribute of graph in _init_all_nodes_set?

**Comment by Skeleton003** [comment]
> > > > _init_all_nodes_set
> > > 
> > > 
> > > Why do we need to change _init_all_nodes_set?
> > 
> > 
> > Please take a kind look at the [Description](https://github.com/dmlc/dgl/pull/7147#issue-2151742994). The purpose of this PR is to generate `all_nodes_set` with a specific dtype, and this purpose cannot be achieved if neither `FusedCSCSamplingGraph.num_nodes` nor `OnDiskDataset._init_all_nodes_set` is changed, because currently `FusedCSCSamplingGraph.num_nodes` always returns `Union[int, Dict[str, int]]` and `ItemSet` cannot judge the dtype of an `int`.
> 
> How the dtype should be decided? By external users or depend on any attribute of graph in _init_all_nodes_set?

Same as `graph.indices.dtype`.

**Comment by mfbalin** [comment]
> @mfbalin please take a look at https://github.com/dmlc/dgl/blob/master/python/dgl/graphbolt/impl/ondisk_dataset.py#L824. OnDiskDataset's `_init_all_nodes_set` makes use of `num_nodes` API of `FusedCSCSamplingGraph` to generate `all_nodes_set`. Could you please also modify the return type of `FusedCSCSamplingGraph.num_nodes` from `Union[int, Dict[str, int]]` to `Union[torch.Tensor, Dict[str, torch.Tensor]]` and test if your changes work correctly? Many thanks.

@Skeleton003 I think this modification needs to be done in #7127. Otherwise, I will have to re-do all the changes in your PR here. Can you incorporate the `_init_all_nodes_set` modification in #7127?

**Comment by Skeleton003** [comment]
> > @mfbalin please take a look at https://github.com/dmlc/dgl/blob/master/python/dgl/graphbolt/impl/ondisk_dataset.py#L824. OnDiskDataset's `_init_all_nodes_set` makes use of `num_nodes` API of `FusedCSCSamplingGraph` to generate `all_nodes_set`. Could you please also modify the return type of `FusedCSCSamplingGraph.num_nodes` from `Union[int, Dict[str, int]]` to `Union[torch.Tensor, Dict[str, torch.Tensor]]` and test if your changes work correctly? Many thanks.
> 
> @Skeleton003 I think this modification needs to be done in #7127. Otherwise, I will have to re-do all the changes in your PR here. Can you incorporate the `_init_all_nodes_set` modification in #7127? Or if you wouldn't mind, I can push to your branch there.

Sorry for my late reply due to jet lag. Of course I can modify the `_init_all_nodes_set` in #7127 . thank u

**Comment by mfbalin** [comment]
> > > @mfbalin please take a look at https://github.com/dmlc/dgl/blob/master/python/dgl/graphbolt/impl/ondisk_dataset.py#L824. OnDiskDataset's `_init_all_nodes_set` makes use of `num_nodes` API of `FusedCSCSamplingGraph` to generate `all_nodes_set`. Could you please also modify the return type of `FusedCSCSamplingGraph.num_nodes` from `Union[int, Dict[str, int]]` to `Union[torch.Tensor, Dict[str, torch.Tensor]]` and test if your changes work correctly? Many thanks.
> > 
> > 
> > @Skeleton003 I think this modification needs to be done in #7127. Otherwise, I will have to re-do all the changes in your PR here. Can you incorporate the `_init_all_nodes_set` modification in #7127? Or if you wouldn't mind, I can push to your branch there.
> 
> Sorry for my late reply due to jet lag. Of course I can modify the `_init_all_nodes_set` in #7127 . thank u

Please pardon me, I couldn't wait and pushed my changes to your PR. Hope it is okay.

**Comment by Skeleton003** [comment]
> > > > @mfbalin please take a look at https://github.com/dmlc/dgl/blob/master/python/dgl/graphbolt/impl/ondisk_dataset.py#L824. OnDiskDataset's `_init_all_nodes_set` makes use of `num_nodes` API of `FusedCSCSamplingGraph` to generate `all_nodes_set`. Could you please also modify the return type of `FusedCSCSamplingGraph.num_nodes` from `Union[int, Dict[str, int]]` to `Union[torch.Tensor, Dict[str, torch.Tensor]]` and test if your changes work correctly? Many thanks.
> > > 
> > > 
> > > @Skeleton003 I think this modification needs to be done in #7127. Otherwise, I will have to re-do all the changes in your PR here. Can you incorporate the `_init_all_nodes_set` modification in #7127? Or if you wouldn't mind, I can push to your branch there.
> > 
> > 
> > Sorry for my late reply due to jet lag. Of course I can modify the `_init_all_nodes_set` in #7127 . thank u
> 
> @Skeleton003 Please pardon me, I couldn't wait and pushed my changes to your PR. Hope it is okay.

It's OK! Thank you for fixxing the bugs!

**Pull request merged**